include/zephyr/sys/time_units.h

```diff
@@ -111,70 +111,128 @@ static TIME_CONSTEXPR inline int sys_clock_hw_cycles_per_sec(void)
+	((__to_hz) == (__from_hz))
+
+/* true if the conversion requires a simple integer multiply */
+#define z_tmcvt_mul_ratio(__from_hz, __to_hz) \
+	((__to_hz) > (__from_hz) && (__to_hz) % (__from_hz) == 0U)
+
+/* true if the conversion requires a simple integer division */
+#define z_tmcvt_div_ratio(__from_hz, __to_hz) \
+	((__from_hz) > (__to_hz) && (uint64_t) (__from_hz) % (__to_hz) == 0U)
+
+/*
+ * Compute the offset needed to round the result correctly when
+ * the conversion requires a simple integer division
+ */
+#define z_tmcvt_off_div(__from_hz, __to_hz, __round_up, __round_off)	\
+	((__round_off) ? ((__from_hz) / (__to_hz)) / 2 :		\
+	 (__round_up) ? ((__from_hz) / (__to_hz)) - 1 :			\
+	 0)
+
+/*
+ * Compute the offset needed to round the result correctly when
+ * the conversion requires a full mul/div
+ */
+#define z_tmcvt_off_gen(__from_hz, __to_hz, __round_up, __round_off)	\
+	((__round_off) ? (__from_hz) / 2 :				\
+	 (__round_up) ? (__from_hz) - 1 :				\
+	 0)
+
+/* General 32-bit conversion for constant frequencies */
+#define z_tmcvt_const_32(__t, __from_hz, __to_hz, __round_up, __round_off) \
+	((uint32_t) (((uint64_t) (__t)*(__to_hz) + \
+		      z_tmcvt_off_gen(__from_hz, __to_hz, __round_up, __round_off)) / (__from_hz)))
+
+/* Fast 64-bit conversion for constant frequencies. This relies on the multiply not overflowing */
+#define z_tmcvt_const_64_fast(__t, __from_hz, __to_hz, __round_up, __round_off) \
+	(((uint64_t) (__t)*(__to_hz) + \
+	  z_tmcvt_off_gen(__from_hz, __to_hz, __round_up, __round_off)) / (__from_hz))
+
+/* Slow 64-bit conversion for constant frequencies. This avoids overflowing the multiply */
+#define z_tmcvt_const_64_slow(__t, __from_hz, __to_hz, __round_up, __round_off) \
+	(((uint64_t) (__t) / (__from_hz))*(__to_hz) +			\
+	 (((uint64_t) (__t) % (__from_hz))*(__to_hz) +		\
+	  z_tmcvt_off_gen(__from_hz, __to_hz, __round_up, __round_off)) / (__from_hz))
+
+/* General 64-bit conversion for constant frequencies. Uses one of the two above macros */
+#define z_tmcvt_const_64(__t, __from_hz, __to_hz, __round_up, __round_off) \
+	(Z_TMCVT_USE_FAST_ALGO(__from_hz, __to_hz) ?			\
+	 z_tmcvt_const_64_fast(__t, __from_hz, __to_hz, __round_up, __round_off) : \
+	 z_tmcvt_const_64_slow(__t, __from_hz, __to_hz, __round_up, __round_off))
+
+/* Functions for cases with non-constant frequencies */
+uint32_t _z_tmcvt_32(uint64_t t, uint32_t from_hz,
+		     uint32_t to_hz, bool round_up,
+		     bool round_off);
+
+uint64_t _z_tmcvt_64(uint64_t t, uint32_t from_hz,
+		     uint32_t to_hz, bool round_up,
+		     bool round_off);
+
+/* Convert, generating a 32-bit result */
+#define z_tmcvt_32(__t, __from_hz, __to_hz, __const_hz, __round_up, __round_off) \
+	((__const_hz) ?							\
+	 (								\
+		 z_tmcvt_identity(__from_hz, __to_hz) ?			\
+		 (uint32_t) (__t)					\
+		 :							\
+		 z_tmcvt_div_ratio(__from_hz, __to_hz) ?		\
+		 (							\
+			 (int64_t) (__t) < (int64_t) BIT64(32) -	\
```

Comment:
This is really hard to follow now. The original approach was much easier to read. Don't have an immediate idea how to fix it but looking again might be worth it?

I think the general approach of having two kinds of macros per conversion algorithm is the best way to go:
- one macro that checks applicability of a certain algorithm
- one macro that applies the algorithm

These macros can be named and placed in pairs such that their correspondence is obvious and easy to follow.

Then the basic functional approach to the conversion is `is_a_applicable(...) ? convert_by_a() : is_b_applicable(...) : convert_by_b() ... : convert_by_n()`.

This was the idea all the time anyway but now it can be made obvious and trivial rather than trying to be overly clever. Just my personal taste of course...

Reply:
Great idea. I've implemented this and the resulting `z_tmcvt_32` and `z_tmcvt_64` macros are pretty simple looking now.